In [1]:
import pandas as pd

In [22]:
xls = pd.ExcelFile('KPMG_VI_New_raw_data_update_final.xlsx')
Transactions = pd.read_excel(xls, 'Transactions',skiprows=1)
NewCustomerList = pd.read_excel(xls, 'NewCustomerList',skiprows=1)
CustomerDemographic = pd.read_excel(xls, 'CustomerDemographic',skiprows=1)
CustomerAddress = pd.read_excel(xls, 'CustomerAddress',skiprows=1)

# Transactions

In [23]:
Transactions.head(3)

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   

  order_status          brand product_line product_class product_size  \
0     Approved          Solex     Standard        medium       medium   
1     Approved  Trek Bicycles     Standard        medium        large   
2     Approved     OHM Cycles     Standard           low       medium   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0

## Finding out Null values 

In [107]:
#function for finding out features containing null values
def null_features(df):
    null=[]
    for feature in df.isnull().sum().index:
        if df.isnull().sum()[feature]!=0:
            null.append(feature)
    return null

In [106]:
null_features(Transactions)

['online_order',
 'brand',
 'product_line',
 'product_class',
 'product_size',
 'standard_cost',
 'product_first_sold_date']

In [152]:
Transactions.isnull().sum()/len(Transactions)*100 #percent of missing values

transaction_id             0.000
product_id                 0.000
customer_id                0.000
transaction_date           0.000
online_order               1.800
order_status               0.000
brand                      0.985
product_line               0.985
product_class              0.985
product_size               0.985
list_price                 0.000
standard_cost              0.985
product_first_sold_date    0.985
dtype: float64

In [28]:
Transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [162]:
len(Transactions.customer_id.unique())

3494

In [34]:
Transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

# NewCustomerList

In [81]:
NewCustomerList.head(3)

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   

         DOB               job_title job_industry_category     wealth_segment  \
0 1957-07-12         General Manager         Manufacturing      Mass Customer   
1 1970-03-22     Structural Engineer              Property      Mass Customer   
2 1974-08-28  Senior Cost Accountant    Financial Services  Affluent Customer   

  deceased_indicator owns_car  ...  state    country  property_valuation  \
0                  N      Yes  ...    QLD  Australia                   6   
1                  N       No  ...    NSW  Australia                  11   
2                  N       No  ...    VIC  Australia                   5   

  Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  Rank  \
0        0.56        0.70       0.8750     0.743750            1     1   
1        0.89        0.89       1.1125     0.945625            1     1   
2        1.01        1.01       1.0100     1.010000            1     1   

     Value  
0  1.71875  
1  1.71875  
2  1.71875  

[3 rows x 23 columns]

In [89]:
NewCustomerList.columns[16]

'Unnamed: 16'

In [108]:
null_features(NewCustomerList)

['last_name', 'DOB', 'job_title', 'job_industry_category']

In [153]:
NewCustomerList.isnull().sum()/len(NewCustomerList)*100

first_name                              0.0
last_name                               2.9
gender                                  0.0
past_3_years_bike_related_purchases     0.0
DOB                                     1.7
job_title                              10.6
job_industry_category                  16.5
wealth_segment                          0.0
deceased_indicator                      0.0
owns_car                                0.0
tenure                                  0.0
address                                 0.0
postcode                                0.0
state                                   0.0
country                                 0.0
property_valuation                      0.0
Unnamed: 16                             0.0
Unnamed: 17                             0.0
Unnamed: 18                             0.0
Unnamed: 19                             0.0
Unnamed: 20                             0.0
Rank                                    0.0
Value                           

In [124]:
NewCustomerList.Value.value_counts()

0.637500    13
1.062500    12
1.237500     9
0.892500     9
0.945625     9
            ..
0.630000     1
0.340000     1
0.875500     1
0.867000     1
1.142187     1
Name: Value, Length: 324, dtype: int64

In [123]:
len(NewCustomerList.Rank.unique())

324

In [83]:
NewCustomerList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [84]:
NewCustomerList.describe()

past_3_years_bike_related_purchases       tenure     postcode  \
count                          1000.000000  1000.000000  1000.000000   
mean                             49.836000    11.388000  3019.227000   
std                              27.796686     5.037145   848.895767   
min                               0.000000     0.000000  2000.000000   
25%                              26.750000     7.000000  2209.000000   
50%                              51.000000    11.000000  2800.000000   
75%                              72.000000    15.000000  3845.500000   
max                              99.000000    22.000000  4879.000000   

       property_valuation  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
count         1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean             7.397000     0.747340     0.839005     0.942673     0.870514   
std              2.758804     0.205082     0.248858     0.294832     0.280891   
min              1.000000     0.400000     0.400000     0.400000     0.340000   
25%              6.000000     0.570000     0.637500     0.712500     0.658750   
50%              8.000000     0.750000     0.820000     0.912500     0.842625   
75%              9.000000     0.920000     1.031875     1.142969     1.062500   
max             12.000000     1.100000     1.375000     1.718750     1.718750   

       Unnamed: 20         Rank        Value  
count  1000.000000  1000.000000  1000.000000  
mean    498.819000   498.819000     0.881714  
std     288.810997   288.810997     0.293525  
min       1.000000     1.000000     0.340000  
25%     250.000000   250.000000     0.649531  
50%     500.000000   500.000000     0.860000  
75%     750.250000   750.250000     1.075000  
max    1000.000000  1000.000000     1.718750

# CustomerDemographic

In [125]:
CustomerDemographic.head(3)

customer_id first_name  last_name gender  \
0            1    Laraine  Medendorp      F   
1            2        Eli    Bockman   Male   
2            3      Arlin     Dearle   Male   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   

  job_industry_category wealth_segment deceased_indicator  \
0                Health  Mass Customer                  N   
1    Financial Services  Mass Customer                  N   
2              Property  Mass Customer                  N   

                        default owns_car  tenure  
0                            "'      Yes    11.0  
1  <script>alert('hi')</script>      Yes    16.0  
2           2018-02-01 00:00:00      Yes    15.0

In [126]:
null_features(CustomerDemographic)

['last_name', 'DOB', 'job_title', 'job_industry_category', 'default', 'tenure']

In [164]:
len(CustomerDemographic.customer_id.unique())

4000

In [127]:
CustomerDemographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [130]:
CustomerDemographic.gender.value_counts()

Female    2037
Male      1872
U           88
Femal        1
M            1
F            1
Name: gender, dtype: int64

In [137]:
CustomerDemographic.deceased_indicator.value_counts()

N    3998
Y       2
Name: deceased_indicator, dtype: int64

In [138]:
CustomerDemographic.default.value_counts()

100                                       113
1                                         112
-1                                        111
-100                                       99
١٢٣                                     53
                                         ... 
社æç§å­¸é¢èªå­¸ç ç©¶æ                     31
/dev/null; touch /tmp/blns.fail ; echo     30
âªâªtest⪠                                29
ì¸ëë°í ë¥´                                 27
,ãã»:*:ã»ãâ( â» Ï â» )ãã»:*:ã»ãâ           25
Name: default, Length: 90, dtype: int64

In [128]:
CustomerDemographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

# CustomerAddress

In [141]:
CustomerAddress.head(3)

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9

In [142]:
null_features(CustomerAddress)

[]

In [143]:
CustomerAddress.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [149]:
CustomerAddress.property_valuation.value_counts()

9     647
8     646
10    577
7     493
11    281
6     238
5     225
4     214
12    195
3     186
1     154
2     143
Name: property_valuation, dtype: int64

In [144]:
CustomerAddress.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

## Checking duplicated rows 

In [75]:
def duplicated_rows(df):
    duplicate = df[df.duplicated()]
    if len(duplicate)==0:
        return 0
    else:
        return duplicate

In [76]:
duplicated_rows(Transactions)

0

In [77]:
duplicated_rows(NewCustomerList)

0

In [79]:
duplicated_rows(CustomerDemographic)

0

In [80]:
duplicated_rows(CustomerAddress)

0

# Inferences

1)	Transactions:


•	online_order, brand, product_line, product_class, product_size, standard_cost, product_first_sold_date- has null values (approximately 1%), we may try imputation techniques to replace the null values.

•	product_first_sold_date - Datatype need to be converted into datetime, but is in float at present. It can be converted into datetime format (UTC). Please confirm.



2)	NewCustomerList:

•	Column no. 16, 17, 18, 19, 20 are missing. Please update me on the same.

•	'last_name', 'DOB', 'job_title'(10.6%), 'job_industry_category'(16.5%)- has null values which can be taken care of applying imputation techniques.

•	‘deceased_indicator’ is a constant feature. Is it an important feature to consider?

•	property_valuation - Unit not mentioned. Please confirm.

•	Value- Unit not mentioned. Please confirm.


3)	CustomerDemographic:

•	Gender- has typing mistakes

•	Default- has many inconsistent/garbage values


4)	CustomerAddress:

•	Country- is a constant feature

•	property_valuation- Unit not mentioned


In [159]:
import datetime
your_timestamp = 36361.0
date = datetime.datetime.utcfromtimestamp(your_timestamp)
date

datetime.datetime(1970, 1, 1, 10, 6, 1)

_No duplicate entries found_